cunumeric/array.py

```diff
@@ -2227,19 +2290,22 @@ def _diag_helper(
                     raise ValueError("output array has wrong shape")
                 a = a._maybe_convert(out.dtype, (a, out))
             else:
-                out = ndarray(
-                    shape=out_shape, dtype=self.dtype, inputs=(self,)
-                )
-            if out is None and dtype is not None:
-                a = a._maybe_convert(dtype, (a,))
+                if dtype is not None:
+                    a = a._maybe_convert(dtype, (a,))
+                out = ndarray(shape=out_shape, dtype=a.dtype, inputs=(self,))
```

Comment:
fix for unreachable code (`out` was never None in the old code) cc @ipdemes

Reply:
Took this as an opportunity to update the handling of `dtype` in this function, to match broader NumPy practice (`dtype` if present sets the precision of the computation, `out.dtype` is used as `dtype` if it's not present, computation result is cast if necessary to `out.dtype` if `out` is present).